### ISSUE TITLE
Why using offset in calculating eigenvector?

### ISSUE THREAD

**Issue by cszachary** [loukasa/graph-coarsening#1]
Hi, 

Thanks for sharing your code. 

I just wonder why you use offset in the following code. 

     if 'variation' in method:
            
            if level == 1:
                if (Uk is not None) and (lk is not None) and (len(lk)>= K):
                    mask = lk<1e-10; lk[mask] = 1; lsinv = lk**(-0.5); lsinv[mask] = 0
                    B = Uk[:,:K] @ np.diag(lsinv[:K])    
                else:
                    offset = 2*max(G.dw)
                    T = offset*sp.sparse.eye(G.N, format='csc') - G.L
                    lk, Uk = sp.sparse.linalg.eigsh(T, k=K, which='LM', tol=1e-5)
                    lk = (offset-lk)[::-1]
                    Uk = Uk[:,::-1]                
                    mask = lk<1e-10; lk[mask] = 1; lsinv = lk**(-0.5); lsinv[mask] = 0
                    B = Uk @ np.diag(lsinv)    
According to the experiment results, a larger offset value will result in a smaller $\epsilon$ error bound and vice versa. Is the offset value a hyper-parameter? If yes, how should I choose it?

**Comment by loukasa** [comment]
Hello,

The offset trick allows one to use the Lanczos method, which finds the
eigenvalues of largest magnitude, in order to find the eigenvalues of
smallest magnitude. Normally, the Lanczos implementation should do the same
internally (if the correct flag is set), but in my tests my method was
always faster.

Actually, the trick should work when the offset is close to and larger than
the norm of L. To avoid actually computing the norm, I use the fact that
||L||_2 <= 2 max_degree, which is a well known upper bound.

In summary, changing the offset is not advised.

**Comment by cszachary** [comment]
Thanks. It's much more clear to me now.
